File: app/src/main/java/se/gustavkarlsson/skylight/android/dagger/modules/definitive/LatestAuroraReportObservableModule.java
package se.gustavkarlsson.skylight.android.dagger.modules.definitive;

import org.threeten.bp.Clock;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import se.gustavkarlsson.skylight.android.cache.SingletonCache;
import se.gustavkarlsson.skylight.android.models.AuroraFactors;
import se.gustavkarlsson.skylight.android.models.AuroraReport;
import se.gustavkarlsson.skylight.android.models.factors.Darkness;
import se.gustavkarlsson.skylight.android.models.factors.GeomagActivity;
import se.gustavkarlsson.skylight.android.models.factors.GeomagLocation;
import se.gustavkarlsson.skylight.android.models.factors.Visibility;
import se.gustavkarlsson.skylight.android.observers.ObservableData;

import static se.gustavkarlsson.skylight.android.dagger.Names.LATEST_NAME;

@Module
public abstract class LatestAuroraReportObservableModule {

	@Provides
	@Singleton
	@Named(LATEST_NAME)
	static ObservableData<AuroraReport> provideLatestAuroraReportObservable(@Named(LATEST_NAME) SingletonCache<AuroraReport> cache, Clock clock) {
		AuroraReport report = cache.getValue();
		if (report == null) {
			AuroraFactors factors = new AuroraFactors(
					new GeomagActivity(null),
					new GeomagLocation(null),
					new Darkness(null),
					new Visibility(null)
			);
			report = new AuroraReport(clock.millis(), null, factors);
		}
		return new ObservableData<>(report);
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/dagger/modules/definitive/LatestAuroraReportObservableModule.kt
package se.gustavkarlsson.skylight.android.dagger.modules.definitive

import dagger.Module
import dagger.Provides
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.Subject
import org.threeten.bp.Clock
import se.gustavkarlsson.skylight.android.cache.SingletonCache
import se.gustavkarlsson.skylight.android.dagger.LATEST_NAME
import se.gustavkarlsson.skylight.android.entities.*
import se.gustavkarlsson.skylight.android.extensions.now
import javax.inject.Named
import javax.inject.Singleton

@Module
class LatestAuroraReportObservableModule {

	@Provides
	@Singleton
	@Named(LATEST_NAME)
	fun provideLatestAuroraReportObservable(@Named(LATEST_NAME) subject: Subject<AuroraReport>): Observable<AuroraReport> {
		return subject
	}

	@Provides
	@Singleton
	@Named(LATEST_NAME)
	fun provideLatestAuroraReportObserver(@Named(LATEST_NAME) subject: Subject<AuroraReport>): Observer<AuroraReport> {
		return subject
	}

	@Provides
	@Singleton
	@Named(LATEST_NAME)
	fun provideLatestAuroraReportSubject(@Named(LATEST_NAME) cache: SingletonCache<AuroraReport>, clock: Clock): Subject<AuroraReport> {
		var report = cache.value
		if (report == null) {
			val factors = AuroraFactors(
				GeomagActivity(null),
				GeomagLocation(null),
				Darkness(null),
				Visibility(null)
			)
			report = AuroraReport(clock.now, null, factors)
		}
		val subject = BehaviorSubject.createDefault(report)
		subject.doOnNext { cache.value = it }
		return subject
	}
}
